api/controllers/project.js

```diff
@@ -756,6 +760,132 @@ exports.protectedUnPublishPin = async function (args, res) {
   }
 };
 
+exports.publicCACSignUp = async function ( args, res) {
+  // sign this user up for CAC on the project.
+
+  const CACUser = mongoose.model('CACUser');
+  const Project = mongoose.model('Project');
+
+  // Clear out anything dangerous first, before creating an instance of the user
+  let cacObject     = args.swagger.params.cac.value;
+  const projectId   = args.swagger.params.projId.value;
+
+  delete cacObject.read;
+  delete cacObject.write;
+
+  // Tie this instance of user to a project
+  cacObject.project = projectId;
+
+  let cacUserToAdd  = new CACUser(cacObject);
+
+  // Find this email address in the caclist.
+  let cacUser = await CACUser.findOne({
+    _schemaName: 'CACUser',
+    email: cacUserToAdd.email,
+    project: mongoose.Types.ObjectId(projectId)
+  });
+
+  if (!cacUser) {
+    // Not found, create the cacUser object in the project.
+    cacUserToAdd.read.push('sysadmin');
+    cacUserToAdd.write.push('sysadmin');
+    cacUser = await cacUserToAdd.save();
+  }
+
+  // AddToSet this if it isn't already in the cac data.
+  try {
+    await Project.findOneAndUpdate(
+      { _id: mongoose.Types.ObjectId(projectId) },
+      {
+        $addToSet: { "cacMembers": mongoose.Types.ObjectId(cacUser._id) }
+      },
+      { new: true }
+    );
+    Utils.recordAction('Post', 'ProjectCACMember', 'public', cacUser._id);
+    // We don't want to return anything but a 200 OK.
+    return Actions.sendResponse(res, 200, {});
+  } catch (e) {
+    defaultLog.info("Error inserting user into the project cac:", e);
+    return Actions.sendResponse(res, 500, {});
+  }
+};
+
+exports.protectedCACRemoveMember = async function (args, res) {
+  // Remove a user from CAC
+  const projId = args.swagger.params.projId.value;
+  const member = args.swagger.params.member.value;
+  defaultLog.info("Delete CAC Member:", member, " from Project:", projId);
+
+  const Project = mongoose.model('Project');
+  const CACUser = mongoose.model('CACUser');
+  try {
+    // First remove it from the CACUser collection
+    await CACUser.remove({_id: mongoose.Types.ObjectId(member._id)});
```

Comment:
Just wondering if these two operations should change order. That way, if the second action fails it doesn't leave a user ID associated to a project that doesn't exist.

Reply:
Hmm.. maybe better to try catch each independently?

Reply:
Could be. I think as long as it solves the issue of the project pointing to non-existent users. If there are two try/catches I think it would still want to reverse the call order

Reply:
I see what you're saying - good idea.  I'll reverse.

Reply:
Fixed in 3c65c7e.